jumanji/environments/routing/macvrp/types.py

```diff
@@ -0,0 +1,142 @@
+        this time, a penalty (based on early coeff) will be handed out.
+    end: this array stores the latest times that the customers
+        want their packages to be picked up. If a vehicle arrives after
+        this time, a penalty (based on late coeff) will be handed out.
+    """
+
+    start: chex.Array  # Shape: (num_customers,)
+    end: chex.Array  # Shape: (num_customers,)
+
+
+@dataclass
+class PenalityCoeff:
+    """Tree structure that represents the penalty coefficients.
+    early: this array stores the early penalty coefficient values for
+        each customer. If a vehicle arrives before the customer's window
+        start time it receives
+        pentality = (TimeWindow.start - Vehicle.local_times) * PenalityCoeff.early.
+    late: this array stores the late penalty coefficient values for
+        each customer. If a vehicle arrives after the customer's window
+        end time it receives
+        pentality = (Vehicle.local_times - TimeWindow.end) * PenalityCoeff.late.
+    """
+
+    early: chex.Array  # Shape: (num_customers,)
+    late: chex.Array  # Shape: (num_customers,)
+
+
+@dataclass
+class Vehicle:
+    """Vehicle tree structure.
+    local_times: This array stores the current local times of
+        each vehicle. This is necessary as each vehicle has traveled
+        a different total distance and therefore has different
+        local times for the same environment step.
+    positions: This array stores the positions (locations) of each of
+        the vehicles. Here 0 means that the vehicle is at the DEPOT and 1+ that
+        it is at a customer.
+    capacities: This array stores the capacities of each vehicle. This
+        represents the total number of packages/volume that the vehicle can still
+        add before needing to return to the depot.
+    distances: This array stores the total distances traveled by each vehicle
+        thus far.
+    time_penalties: This array stores the total time penalties that each vehicle
+        has received thus far.
+    """
+
+    local_times: chex.Array  # Shape: (num_vehicles,)
+    positions: chex.Array  # Shape: (num_vehicles,)
+    capacities: chex.Array  # Shape: (num_vehicles,)
+    distances: chex.Array = None  # Shape: (num_vehicles,)
+    time_penalties: chex.Array = None  # Shape: (num_vehicles,)
+
+
+@dataclass
+class State:
+    """
+    nodes: Customer node coordinates and demands.
+    windows: Time windows within which a customer should be visited.
+    coeffs: Coefficient values used to calculate penalties if the vehicles arrive
+        outside their respective time windows.
+    vehicles: General information related to each vehicle.
+    order: This array stores the history of each vehicle by tracking what customer each
+        vehicle was at each environment step. This is used for rendering.
+    step_count: The current step count in the environment.
+    key: Unused key.
```

Comment:
This is extremely important. Cf [previous bug](https://github.com/instadeepai/jumanji-internal/pull/490) that got fixed.
```suggestion
    key: random key used for auto-reset.
```

Reply:
Updated.